### PR TITLE
EVM: Block template implementation and validation optimization

### PR DESCRIPTION
**Pull request by sieniven**
### Block template implementation
- This PR implements block templates from the old txqueue, and refactors validation and block construction pipeline to optimize EVM executions.
- Block templates replaces the old transaction queues - create new block and connect block pipeline creates a block template and adds / remove transactions from the template.
- We now keep track of state root, receipts, and logs inside each TemplateTxItem.
- We also pass the EVM beneficiary to the block template vicinity, which the old txqueue did not hold which could have led to state mismatches on construct block.
- Construct block pipeline will now no longer redundantly re-execute all txs again on the EVM, optimizing the validation pipeline.
- failed_transactions pipeline inside construct block is discontinued since that is redundant with the block template design.
- New construct block pipeline constructs the block from the block template instance by:
   - Consolidates all transactions and receipts
   - Calculating the total transaction fees, total burnt fees and total priority fees
   - Pays the miner
   - Block construction
- The PR improves node performance by >50%.
- Note that all of the validation changes in the block template design refactor are backwards compatible, and **should not change anything related to current consensus pipeline**.

### Bug fixes
- Fixes genesis block bug from previously appending DST20 migration txs at the end of genesis block to appending the DST20 migration txs after the contract deployment txs.
- Fixes block vicinity in backend state to pass block base fees and block difficulty into the EVM when executing txs
- Fixes all backend vicinity state passed in eth calls and EVM executions to be explicit to ensure desired functionalities on transaction executions
- These changes are **not backwards compatible**.

## Implications

- Storage
  - [ ] Database reindex required
  - [ ] Database reindex optional
  - [ ] Database reindex not required
  - [x] None

- Consensus
  - [x] Network upgrade required
  - [ ] Includes backward compatible changes
  - [ ] Includes consensus workarounds
  - [ ] Includes consensus refactors
  - [ ] None

### CODE REVIEW

File: lib/ain-evm/src/evm.rs [hunk 1]
```diff
@@ -442,77 +272,236 @@ impl EVMServices {
...
-            },
+            template.vicinity.clone(),
         )?;
+        let mut executor = AinExecutor::new(&mut backend);
 
+        let (parent_hash, _) = self
+            .block
+            .get_latest_block_hash_and_number()?
+            .unwrap_or_default(); // Safe since calculate_base_fee will default to INITIAL_BASE_FEE
+        let base_fee = self.block.calculate_base_fee(parent_hash)?;
```
Comment (Jouzo) [inline lib/ain-evm/src/evm.rs:308]: We can store base_fee in block template as well

Comment (sieniven) [reply lib/ain-evm/src/evm.rs:306]: Nice spot, saw it as well, fixed in commit ded07f5 which also adds block difficulty into the template.

File: lib/ain-evm/src/blocktemplate.rs [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+    ///
+    /// # Safety
+    ///
+    /// Result cannot be used safety unless cs_main lock is taken on C++ side
+    /// across all usages. Note: To be replaced with a proper lock flow later.
+    ///
+    pub unsafe fn get_txs_cloned_in(&self, template_id: u64) -> Result<Vec<TemplateTxItem>> {
+        self.with_block_template(template_id, BlockTemplate::get_cloned_transactions)
+    }
```
Comment (Jouzo) [inline lib/ain-evm/src/blocktemplate.rs:146]: Unused, can remove

File: src/dfi/govvariables/attributes.h [hunk 3]
```diff
@@ -588,8 +588,8 @@ class ATTRIBUTES : public GovVariable, public AutoRegistrator<GovVariable, ATTRI
     static const std::map<std::string, uint8_t> &allowedVaultIDs();
     static const std::map<std::string, uint8_t> &allowedRulesIDs();
     static const std::map<uint8_t, std::map<std::string, uint8_t>> &allowedKeys();
-    static const std::map<uint8_t, std::map<uint8_t, std::function<ResVal<CAttributeValue>(const std::string &)>>>
-        &parseValue();
+    static const std::map<uint8_t, std::map<uint8_t, std::function<ResVal<CAttributeValue>(const std::string &)>>> &
```
Comment (Bushstar) [inline src/dfi/govvariables/attributes.h:591]: Is this how the new clang format looks?

**Review by Bushstar — DISMISSED** [review]

File: src/dfi/govvariables/attributes.h (hunk 3 above)
Comment (sieniven) [reply src/dfi/govvariables/attributes.h:591]: I have no idea 🥲

File: lib/ain-evm/src/blocktemplate.rs (hunk 2 above)
Comment (sieniven) [reply lib/ain-evm/src/blocktemplate.rs]: Resolved in commit [ef93279](https://github.com/DeFiCh/ain/pull/2566/commits/ef93279ac70962c9899eb47e0d00c8bbf559153b)

File: src/dfi/govvariables/attributes.h (hunk 3 above)
Comment (prasannavl) [reply src/dfi/govvariables/attributes.h:591]: Weird. Possible. Not sure if there's a config.

**Review by Jouzo — DISMISSED** [review]

File: src/dfi/govvariables/attributes.h (hunk 3 above)
Comment (prasannavl) [reply src/dfi/govvariables/attributes.h]: clang format does push it to the next line :)

Comment (prasannavl) [reply src/dfi/govvariables/attributes.h]: https://github.com/DeFiCh/ain/pull/2566/commits/fd0fcd488610bbba4267c52a7dcc5a9eb40ff9d4

**Pull request merged**